libs/@local/graph/types/typescript/src/validation.ts

```diff
@@ -0,0 +1,235 @@
+      type: "missing";
+    }
+>;
+
+export type JsonSchemaValueTypeMismatch = Omit<
+  JsonSchemaValueTypeMismatchGraphApi,
+  "actual" | "expected"
+> & {
+  actual: [VersionedUrl, ...VersionedUrl[]];
+  expected: [VersionedUrl, ...VersionedUrl[]];
+};
+
+export type PropertyValidationReport = Subtype<
+  PropertyValidationReportGraphApi,
+  | {
+      // The property value validation failed
+      type: "value";
+      /*
+       * Validation for each constraint in the `oneOf` field. The validation is assumed to pass if exactly one of the
+       * constraints passes. In this case, this field will be omitted. Whenever this field is present it can be assumed
+       * that the validation failed.
+       */
+      validations?: (
+        | { status: "success" }
+        | { status: "failure"; data: PropertyValueValidationReport }
+      )[];
+      // It was not possible to infer the correct data type ID for the property.
+      dataTypeInference?: DataTypeInferenceError[];
+      // Converting the data type from `originalDataTypeId` to `dataTypeId` failed
+      valueConversion?: DataTypeConversionErrorGraphApi;
+      // It was not possible to convert the value to its canonical forms.
+      canonicalValue?: DataTypeCanonicalCalculationGraphApi[];
+    }
+  | {
+      // The property array validation failed
+      type: "array";
+      /*
+       * Validation for each constraint in the `oneOf` field. The validation is assumed to pass if exactly one of the
+       * constraints passes. In this case, this field will be omitted. Whenever this field is present it can be assumed
+       * that the validation failed.
+       */
+      validations?: (
+        | { status: "success" }
+        | { status: "failure"; data: PropertyArrayValidationReport }
+      )[];
+    }
+  | {
+      // The property object validation failed
+      type: "object";
+      /*
+       * Validation for each constraint in the `oneOf` field. The validation is assumed to pass if exactly one of the
+       * constraints passes. In this case, this field will be omitted. Whenever this field is present it can be assumed
+       * that the validation failed.
+       */
+      validations?: (
+        | { status: "success" }
+        | { status: "failure"; data: PropertyObjectValidationReport }
+      )[];
+    }
+>;
+
+export type PropertyValueValidationReport = Subtype<
+  PropertyValueValidationReportGraphApi,
+  | {
+      type: "wrongType";
+      data: PropertyValueTypeMismatchGraphApi;
+    }
+  | {
+      type: "valueValidation";
+      data: ValueValidationReport;
+    }
+>;
+
+export type ValueValidationReport = Omit<
+  ValueValidationReportGraphApi,
+  "actual" | "desired" | "abstract" | "incompatible"
+> & {
+  // The actual value has a validation error
+  actual?: ValueValidationError;
+  // The value could not be validated against the data type specified in the schema
+  desired?: ValueValidationError;
```

Comment:
I slightly prefer `provided` over `actual` because I think it makes it clear that it's what the caller has provided – rather than 'actual' which could mean 'actually provided' or 'actually in the schema'.

Similar for 'desired' – is it desired by the caller or the schema? Although feel less strongly about this one.

```suggestion
  provided?: ValueValidationError;
  // The value could not be validated against the data type specified in the schema
  target?: ValueValidationError;
```

Reply:
[`8fa0baa` (#5850)](https://github.com/hashintel/hash/pull/5850/commits/8fa0baa5fa34141623ff8e99f5733f3ad388bf91)